jrnl/output.py

```diff
@@ -25,17 +25,30 @@ def deprecated_cmd(old_cmd, new_cmd, callback=None, **kwargs):
         callback(**kwargs)
 
 
-def list_journals(configuration):
+def list_journals(configuration, format=None):
     from jrnl import config
 
     """List the journals specified in the configuration file"""
-    result = f"Journals defined in config ({config.get_config_path()})\n"
-    ml = min(max(len(k) for k in configuration["journals"]), 20)
-    for journal, cfg in configuration["journals"].items():
-        result += " * {:{}} -> {}\n".format(
-            journal, ml, cfg["journal"] if isinstance(cfg, dict) else cfg
-        )
-    return result
+    if format == "json":
+        import json
+
+        return json.dumps(configuration["journals"])
+    elif format == "yaml":
+        from io import StringIO
+
+        from ruamel.yaml import YAML
+
+        output = StringIO()
+        YAML().dump(configuration["journals"], output)
+        return output.getvalue()
+    else:
+        result = f"Journals defined in config ({config.get_config_path()})\n"
+        ml = min(max(len(k) for k in configuration["journals"]), 20)
+        for journal, cfg in configuration["journals"].items():
+            result += " * {:{}} -> {}\n".format(
+                journal, ml, cfg["journal"] if isinstance(cfg, dict) else cfg
+            )
+        return result
```

Comment:
I see the value in the simplicity of this approach, though I would hope to see the same information regardless of what format is used. In particular, the nested configuration dictionary doesn't flatten the journal paths between journals specified using "journalname: path" and "journal: journalname: path". And while the old approach already combined the concerns of data retrieval and data presentation, it would be good to keep those separate if we're expanding the feature.

Instead, could you try retrieving the journal data first (sort of like the code in the `else` block does), then sending that data to different export functions based on the file format? Probably best as a dictionary object. That way it's all working with the same data while respecting the separation of concerns.

Reply:
So something like this with a different function for each output format?

```python
journal_list = {
  "config_path" : config.get_config_path(),
  "journals" : configuration["journals"],
}

if format == "json":
  return journal_list_to_json(journal_list)
elif format == "yaml":
  return journal_list_to_yaml(journal_list)
else:
  return journal_list_to_stdout(journal_list)
```